### DISCUSSION TITLE
Lazy query limitations

### DISCUSSION THREAD

**Discussion by JoshuaLevett** [ravendb/ravendb#14436] [Q&A] [answered]
I have a situation where I'm trying to fetch a summary of several items at once.
I have series (a document), which have instances (separate documents) that are associated to them.

My issue is that I'm trying to create a summary page, where I want to get the next (upcoming) instance for each series, my first thought was to set up a lazy call for each series we are displaying, and send all the requests at once.
However, upon setting up these queries, they only execute one at a time, despite even the explicit use of ExecuteAllPendingLazyOperations.
The image below was taken after a couple of loops through where the values are actually used. The first value is fetched lazily, but each following one makes a request when .Value is called

![image](https://user-images.githubusercontent.com/45205371/175520846-4d1d1ba0-058c-4ebe-baf3-782188c9d0e2.png)

Each query is set up as below
`var query = session.Query<PhysicalEvent>().Where(x => x.SeriesId == Id && x.ActualStart >= now).OrderBy(x => x.ActualStart).Take(1).Lazily();`
All queries are set up within the same session

The only alternative I can see is to make a single call for all series, that groups instances on the SeriesId, though this has the potential to load many sessions that I do not want/need.
**Is there a better way to do this? (I've given some of the main limitations below)**

Restrictions:
I do not know until query time, when the next instance is, as they can be rescheduled, and may not match their initial schedule
There may be many series (over 30), all of which should be displayed at once
I want to limit the amount of data fetched that isn't needed

Raven version is 5.2.3

P.S
Coming up with a title is hard

**Comment by ayende** [comment]
Can you show a sample test showing this scenario? That does _not_ match our expectation and shouldn't be happening.

**Comment by JoshuaLevett** [comment, marked as answer]
Apologies for the (very) slow response.
But following on from creation of unit tests and our investigation, the issue appears to be caused because calling .Value on an evaluated lazy query, will still trigger evaluation for all non-evaluated queries.

In our scenario we had:
 - Setup lazy calls
 - Loop through above calls
    - In the loop, after fetching the Value, set up another lazy call to be resolved in a second set of lazys
 - Loop through the second set of Lazy calls
 - Output result
 
 The issue was predictably intermittent, as only some of the second lazy values actually needed to set up a call to the database (those that didn't set up a further request, performed as expected).
Our solution has been to delay the setup of the second set of lazy calls until all the first set have been evaluated.

My initial expectation would have been that calling .Value on a lazy query that has been fetched would not cause a call to the database, it would instead run only when .Value was called on one that hasn't been.

Edit:
If I could, I would mark Ayende's reply as the answer

**Reply by ayende** [reply]
Yes, any call to `.Value` will "flush" all the pending lazy operations in the session.
If you are operating on a list of lazy operations and then add additional lazies, we'll flush any pending operations there.
You can force that by calling `ExecuteAllPendingLazyOperations` directly.

**Comment by garayx** [comment]
> My initial expectation would have been that calling .Value on a lazy query that has been fetched would not cause a call to the database, it would instead run only when .Value was called on one that hasn't been.

Can you send a sample code which reproduces this ? (see: https://ravendb.net/docs/article-page/5.1/Csharp/start/test-driver)

**Reply by JoshuaLevett** [reply]
Not perfect, but there are some examples below (only example B passes).

Though the unit tests these have been set up in a pointless way (it doesn't solve a problem), it exemplifies the issue we had.
We have something like:
```
List<ISeries> seriesList;
...
var seriesAsListCardLazily = new List<Lazy<Lazy<ListCard>>>();
foreach (var series in seriesList)
{
    seriesAsListCardLazily.Add(series.AsListCardLazily());
}
var lazyListCards = new List<Lazy<ListCard>>();
foreach (var veryLazyCard in seriesAsListCardLazily)
{
    lazyListCards.Add(veryLazyCard.Value);
}
```

However, as part of `veryLazyCard.Value` it will sometimes set up a lazy database call (to get the remaining capacity).
The problem we had was that it was then being resolved when we looped over to the next `veryLazyCard.Value`.

As "seriesList" may contain many implementations of ISeries, it handles the specifics of how to translate into a 'list card'.
Due to the data needed for these list cards (details of the next session as well as its remaining capacity), we need to make (up to) 2 database calls.

Our current solution is to put a third lazy layer in, to prevent the code from setting up further database calls until we have directly accessed all the previously set up calls.
Meaning that series.AsListCardLazily() now returns a `Lazy<Lazy<Lazy<ListCard>>>`



```
[Test]
        public void LazyCall_ExampleA()
        {
            using (var store = GetDocumentStore())
            {
                #region Arrange
                using (var setupSession = store.OpenSession())
                {
                    SomeDoc docZero = new SomeDoc(0);
                    SomeDoc docOne = new SomeDoc(1);
                    setupSession.Store(docZero);
                    setupSession.Store(docOne);
                    setupSession.SaveChanges();
                }

                int requestsForSecondLazy = 0;
                #endregion

                #region Act
                using (var actSession = store.OpenSession())
                {
                    var lazyA = actSession.Query<SomeDoc>().Where(x => x.Number == 0).Lazily();
                    var lazyB = actSession.Query<SomeDoc>().Where(x => x.Number == 5).Lazily();
                    var requestA = lazyA.Value;

                    var lazyC = actSession.Query<SomeDoc>().Where(x => x.Number == 1).Lazily();
                    var requestB = lazyB.Value;

                    int preDocOneValueRequests = actSession.Advanced.NumberOfRequests;
                    var requestC = lazyC.Value;
                    requestsForSecondLazy = actSession.Advanced.NumberOfRequests - preDocOneValueRequests;
                }
                #endregion

                #region Assert
                Assert.That(requestsForSecondLazy, Is.EqualTo(1));
                #endregion
            }
        }

        [Test]
        public void LazyCall_ExampleB()
        {
            using (var store = GetDocumentStore())
            {
                #region Arrange
                using (var setupSession = store.OpenSession())
                {
                    SomeDoc docZero = new SomeDoc(0);
                    SomeDoc docOne = new SomeDoc(1);
                    setupSession.Store(docZero);
                    setupSession.Store(docOne);
                    setupSession.SaveChanges();
                }

                int requestsForSecondLazy = 0;
                #endregion

                #region Act
                using (var actSession = store.OpenSession())
                {
                    var lazyA = actSession.Query<SomeDoc>().Where(x => x.Number == 0).Lazily();
                    var requestA = lazyA.Value;

                    var lazyC = actSession.Query<SomeDoc>().Where(x => x.Number == 1).Lazily();
                    var requestA_Again = lazyA.Value;

                    int preDocOneValueRequests = actSession.Advanced.NumberOfRequests;
                    var requestC = lazyC.Value;
                    requestsForSecondLazy = actSession.Advanced.NumberOfRequests - preDocOneValueRequests;
                }
                #endregion

                #region Assert
                Assert.That(requestsForSecondLazy, Is.EqualTo(1));
                #endregion
            }
        }

        [Test]
        public void LazyCall_ExampleC()
        {
            using (var store = GetDocumentStore())
            {
                #region Arrange
                using (var setupSession = store.OpenSession())
                {
                    SomeDoc docZero = new SomeDoc(0);
                    SomeDoc docOne = new SomeDoc(1);
                    setupSession.Store(docZero);
                    setupSession.Store(docOne);
                    setupSession.SaveChanges();
                }

                int requestsForSecondLazy = 0;
                #endregion

                #region Act
                using (var actSession = store.OpenSession())
                {
                    var lazyA = actSession.Query<SomeDoc>().Where(x => x.Number == 0).Lazily();
                    var lazyB = actSession.Query<SomeDoc>().Where(x => x.Number == 5).Lazily();

                    actSession.Advanced.Eagerly.ExecuteAllPendingLazyOperations();

                    var lazyC = actSession.Query<SomeDoc>().Where(x => x.Number == 1).Lazily();
                    var requestB = lazyB.Value;

                    int preDocOneValueRequests = actSession.Advanced.NumberOfRequests;
                    var requestC = lazyC.Value;
                    requestsForSecondLazy = actSession.Advanced.NumberOfRequests - preDocOneValueRequests;
                }
                #endregion

                #region Assert
                Assert.That(requestsForSecondLazy, Is.EqualTo(1));
                #endregion
            }
        }




        public class SomeDoc
        {
            public int Number { get; set; }

            public SomeDoc(int number)
            {
                this.Number = number;
            }
        }
```

**Reply by ayende** [reply]
What is `Lazy<Lazy<T>>` here? 

What does `AsListCardLazily` do?

In `var requestA = lazyA.Value;`, that would also `lazyB`
And `lazyC` would be separate, and the call to `lazyB.Value` will likely trigger that.

The lazy code isn't meant to be incrementally built in this manner.

I would expect that `LazyCall_ExampleA` would generate 2 calls to the database. 

And `LazyCall_ExampleB` will generate 3 calls.

Basically, `Lazy.Value` call will trigger eager loading to everything in the session even if the actual value was already loaded.

This is by design

**Reply by JoshuaLevett** [reply]
The purpose of `AsListCardLazily` is to allow implementors to create the queries needed to fetch the content that will be displayed in the card, as it differs between them.
`T` is ultimately what will be displayed / sent to the user, and `Lazy<Lazy<T>>` represents the route from `ISeries` to get there, separated by when lazy calls can be evaluated together.

I didn't mean to imply that this wasn't intentional. When originally raising the ticket, I was unaware of the cause, and the unit tests are setup to represent what my initial assumptions had been.
Every scenario makes 2 calls, the tests are based around when I had (initially) assumed they would occur.

We have resolved our issue, now that we understand what the cause was.

**Reply by ayende** [reply]
The way Lazy works in RavenDB. It aggregates all the still pending operations in the session.
_Any_ first resolution of the a `Lazy` instance will trigger flushing of that to the server.

Lazy<Lazy> doesn't make sense in this model, because as you are building lazy operations while iterating over them, you'll trigger additional calls.

**Reply by ayende** [reply]
You may want to use a _separate_ session, however, that will allow you to scope this.

**Comment by georgiosd** [comment]
If it helps, I always make sure to explicitly call eagerly execute pending
evaluations, before accessing any values


On Wed, 26 Oct 2022 at 13:30, Joshua Levett ***@***.***>
wrote:

> The purpose of AsListCardLazily is to allow implementors to create the
> queries needed to fetch the content that will be displayed in the card.
> T is ultimately what will be displayed / sent to the user, and
> Lazy<Lazy<T>> represents the route from ISeries to get there, separated
> by when lazy calls can be evaluated together.
>
> I didn't mean to imply that this wasn't intentional. When originally
> raising the ticket, I was unaware of the cause, and the unit tests are
> setup to represent what my initial assumptions had been.
> Every scenario makes 2 calls, the tests are based around when I had
> (initially) assumed they would occur.
>
> We have resolved our issue, now that we understand what the cause was.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ravendb/ravendb/discussions/14436#discussioncomment-3968671>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMYG5VMO65RU5IWQZ4AYODWFEB35ANCNFSM5ZXPVPKQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
-- 
Georgios Diamantopoulos
Founder
Zero to MVP
https://zerotomvp.dev
